pkg/vault/contracts/VaultAdmin.sol

```diff
@@ -432,49 +432,104 @@ contract VaultAdmin is IVaultAdmin, VaultCommon, Authentication {
         nonReentrant
         returns (uint256 issuedShares)
     {
+        if (_bufferAssets[wrappedToken] != address(0)) {
+            revert BufferAlreadyInitialized(wrappedToken);
+        }
+
         address underlyingToken = wrappedToken.asset();
 
         if (underlyingToken == address(0)) {
             // Should never happen, but a malicious wrapper could return the zero address and cause the buffer
             // initialization code to run more than once.
-            revert InvalidUnderlyingTokenAsset();
+            revert InvalidUnderlyingToken();
         }
 
+        // Register asset of wrapper, so it cannot change.
+        _bufferAssets[wrappedToken] = underlyingToken;
+
+        // Take debt for initialization assets.
+        _takeDebt(IERC20(underlyingToken), amountUnderlyingRaw);
+        _takeDebt(wrappedToken, amountWrappedRaw);
+
+        // Update buffer balances.
+        bytes32 bufferBalances = _bufferTokenBalances[wrappedToken];
+        bufferBalances = PackedTokenBalance.toPackedBalance(
+            bufferBalances.getBalanceRaw() + amountUnderlyingRaw,
+            bufferBalances.getBalanceDerived() + amountWrappedRaw
+        );
+
+        _bufferTokenBalances[wrappedToken] = bufferBalances;
+
         // Amount of shares to issue is the total underlying token that the user is depositing.
         issuedShares = wrappedToken.convertToAssets(amountWrappedRaw) + amountUnderlyingRaw;
 
-        if (_bufferAssets[wrappedToken] == address(0)) {
-            // Buffer is not initialized yet, so we initialize it.
+        _ensureMinimumTotalSupply(issuedShares);
 
-            // Register asset of wrapper, so it cannot change.
-            _bufferAssets[wrappedToken] = underlyingToken;
+        issuedShares -= _MINIMUM_TOTAL_SUPPLY;
+        _mintBufferShares(wrappedToken, sharesOwner, issuedShares);
+        _mintMinimumBufferSupplyReserve(wrappedToken);
 
-            // Burn MINIMUM_TOTAL_SUPPLY shares, so the buffer can never go back to zero liquidity
-            // (avoids rounding issues with low liquidity).
-            _bufferTotalShares[wrappedToken] = _MINIMUM_TOTAL_SUPPLY;
-            issuedShares -= _MINIMUM_TOTAL_SUPPLY;
-        } else if (_bufferAssets[wrappedToken] != underlyingToken) {
-            // Asset was changed since the first bufferAddLiquidity call.
-            revert WrongWrappedTokenAsset(address(wrappedToken));
-        }
+        emit LiquidityAddedToBuffer(wrappedToken, amountUnderlyingRaw, amountWrappedRaw);
```

Comment:
I prefer the approach to having an add / remove event, but this is at odds to the flow we use in addLiquidity/removeLiquidity where we call the `PoolBalanceChanged` event.